### DISCUSSION TITLE
Parallel Union Find

### DISCUSSION THREAD

**Discussion by erincatto** [jrouwe/JoltPhysics#539] [General]
Hi Jorrit,

I've implemented parallel union find in a Box2D branch. One problem I'm seeing is that contact constraints are added in a non-deterministic order. This means the constraint order changes every time step, reducing stability.

I notice you sort contacts by default for determinism. Did you also notice a stability difference?

Erin

**Comment by jrouwe** [comment]
I did notice a slight difference in behavior if I didn't sort, but it didn't feel less stable in the tests that I ran. The sorting was only meant for determinism.

**Comment by mihe** [comment]
I happened to receive a report about this earlier today, where someone noticed a difference in the stability of stacks depending on whether `mDeterministicSimulation` was on or off, where having it on resulted in better stability.

This isn't exactly the most rigorous test, but you can see this somewhat clearly in this modified version of Jolt's `PyramidTest`:

```cpp
void PyramidTest::Initialize() 
{
	// Floor
	CreateFloor();

	const float cBoxSize = 2.0f;
	const float cBoxSeparation = 0.0f;
	const float cHalfBoxSize = 0.5f * cBoxSize;
	const int cPyramidHeight = 10;

	RefConst<Shape> box_shape = new BoxShape(Vec3::sReplicate(cHalfBoxSize));

	// Pyramid
	for (int i = 0; i < cPyramidHeight - 2; ++i)
		for (int j = i / 2; j < cPyramidHeight - (i + 1) / 2; ++j)
			for (int k = i / 2; k < cPyramidHeight - (i + 1) / 2; ++k)
			{
				RVec3 position(-cPyramidHeight + cBoxSize * j + (i & 1? cHalfBoxSize : 0.0f), 1.0f + (cBoxSize + cBoxSeparation) * i, -cPyramidHeight + cBoxSize * k + (i & 1? cHalfBoxSize : 0.0f));
				BodyCreationSettings box(box_shape, position, Quat::sIdentity(), EMotionType::Dynamic, Layers::MOVING);
				box.mOverrideMassProperties = EOverrideMassProperties::CalculateInertia;
				box.mMassPropertiesOverride.mMass = 1.0f;
				mBodyInterface->CreateAndAddBody(box, EActivation::Activate);
			}

	BodyCreationSettings sphere(new SphereShape(5.0f), RVec3(-cHalfBoxSize, 21.0f, -cHalfBoxSize), Quat::sIdentity(), EMotionType::Dynamic, Layers::MOVING);
	sphere.mOverrideMassProperties = EOverrideMassProperties::CalculateInertia;
	sphere.mMassPropertiesOverride.mMass = 500.0f;
	mBodyInterface->CreateAndAddBody(sphere, EActivation::Activate);
}
```

If you disable the deterministic simulation/sorting while the simulation is running there's a sudden loss of stability.

https://github.com/jrouwe/JoltPhysics/assets/4884246/a1bd094c-959c-443c-ac8b-d198bd8baf99

**Comment by erincatto** [comment]
I'm seeing a big difference for very large stacks. I noticed a decline in stability from Depth-First-Search island building, which tends to sort contacts according to body creation order. I added sorting to union-find to mimic this.

https://github.com/jrouwe/JoltPhysics/assets/7284063/c3c696ba-eaff-4f3c-a929-e75b5797ad81

**Comment by jrouwe** [comment]
Ok, my guess is that the constraint forces are less consistent when the constraints are not sorted so that the warm starting works less well. I did indeed not try this with a stack this large and in @mihe's case there is a heavy body on a stack of lighter bodies which is already difficult for an iterative solver. I am sorting the bodies in the thread that does the actual solving for an island, so that for the common case that there are many small islands, the sort is on a small array and happens in parallel with other islands. For a big stack like this, that obviously doesn't help. I think I'm going to have to add a warning to the ```mDeterministicSimulation``` option (@mihe did you make the default true or false in godot Jolt? - may have to make the default true just like in Jolt itself).

**Reply by mihe** [reply]
I did make `false` the default, since I didn't have any plans on supporting deterministic simulations, but I am actually in the process of pushing 0.1.0-rc3 as we speak, which will remove that project setting from the extension and omit the setting of `mDeterministicSimulation`, making it `true` again. I figured I can always add it back later, once it's more clear to me how I could communicate the trade-offs.
